[Quant][Inductor] Enable quantization linear pattern fusion inside inductor

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106935
* __->__ #106934
* #106782
* #106781
* #105818

**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Xia-Weiwen committed:
[Quant][Inductor] Enable quantization linear pattern fusion inside inductor

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3648,6 +3648,62 @@ def _original_deconv_weight_size(
+    """
+    This function is a helper function to prepare inputs, layout and constant args
+    for linear post-op fusion's create function. The function only supports the CPU device
+    since conv post-op fusion kernel is only supported on CPU right now.
```
```suggestion
    since linear post-op fusion kernel is only supported on CPU right now.
```

Xia-Weiwen replied:
Thanks. It's fixed.

leslie-fang-intel commented on torch/_inductor/ir.py:
```diff
@@ -3648,6 +3648,62 @@ def _original_deconv_weight_size(
+            )
+        output_size = output.size()
+
+        req_stride_order = [0, 1]
```
Why we force un-contiguous input memory format here?

Xia-Weiwen replied:
Thanks. It's fixed.

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/ir.py:
```diff
@@ -4877,6 +4933,117 @@ def create(
         )
 
 
+class QLinearPointwisePT2E(ExternKernelAlloc):
```
if possible, would be great to make this logic more concise, like what you did with quantized_lowerings. If not could we move this and other quantized lowerings to `quantized_ir.py` or something along those lines ? does not have to be in this pr

Xia-Weiwen replied:
Hi @eellison. Thanks for the suggestion. We think it difficult to make it more concise. We will move IR for quantized linear and conv to `quantized_ir.py` in a separate PR. How does that sound to you?

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion inside inductor"


**Summary**
Enable lowering of quantized linear in Inductor

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Xia-Weiwen commented:
@pytorchbot merge
